designs/v1-api.md

```diff
@@ -0,0 +1,357 @@
+
+1. Apply the updated NodePool, NodeClaim, and EC2NodeClass CRDs, which will contain a `v1` versions listed under the `versions` section of the CustomResourceDefinition
+2. Upgrade Karpenter controller to its `v1` version. This version of Karpenter will start reasoning in terms of the `v1` API schema in its API requests. Resources will be converted from the v1beta1 to the v1 version at runtime, using conversion webhooks shipped by the upstream Karpenter project and the Cloud Providers (for NodeClass changes).
+3. Users update their `v1beta1` manifests that they are applying through IaC or GitOps to use the new `v1` version.
+   1. Users that are using multiple NodePools with different kubeletConfigurations that reference the same NodeClasses will need to perform migration from a single NodeClass to multiple NodeClasses. Karpenter v1 [shifts the kubeletConfiguration into the NodeClass API](#moving-spectemplatespeckubelet-into-the-nodeclass) which does not make it possible to represent this many-to-one relationship anymore.
+4. Users remove any conversion annotations from their v1 resources.
+   1. When Karpenter converts from v1beta1 to v1, it maintains round-trippability between the old resource version and the new resource version. To maintain this with schema changes like the `kubeletConfiguration` move, it leverages annotations on the NodePool (e.g. `compatability.karpenter.sh/v1beta1-kubelet-conversion`) to maintain the v1beta1 data. It will be a strict requirement to remove these annotations from the NodePool before upgrading to v1.1.x+.
+5. Karpenter drops the `v1beta1` version from the `CustomResourceDefinition` and the conversion webhooks on the next minor version release of Karpenter, leaving no webhooks and only the `v1` version still present
+
+## NodePool API
+
+```
+apiVersion: karpenter.sh/v1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  template:
+    metadata:
+      labels:
+        billing-team: my-team
+      annotations:
+        example.com/owner: "my-team"
+    spec:
+      nodeClassRef:
+        group: karpenter.k8s.aws # Updated since only a single version will be served
+        kind: EC2NodeClass
+        name: default
+      taints:
+        - key: example.com/special-taint
+          effect: NoSchedule
+      startupTaints:
+        - key: example.com/another-taint
+          effect: NoSchedule
+      requirements:
+        - key: "karpenter.k8s.aws/instance-category"
+          operator: In
+          values: ["c", "m", "r"]
+          minValues: 2 # Alpha field, added for spot best practices support
+      expireAfter: 720h | Never
+      terminationGracePeriod: 1d
+  disruption:
+    budgets:
+      - nodes: 0
+        schedule: "0 10 * * mon-fri"
+        duration: 16h
+        reasons:
```

Comment:
I didn't see disruption by reason section. Should we add that? I saw we did that for termination grace period